src/main/java/edu/cornell/kfs/module/ar/document/web/struts/CuContractsGrantsInvoiceDocumentAction.java

```diff
@@ -41,6 +58,121 @@ public ActionForward prorateBill(ActionMapping mapping, ActionForm form, HttpSer
+            warningMessages.add(configurationService.getPropertyValueAsString(CUKFSKeyConstants.WARNING_CINV_BILLING_PERIOD_END_DATE_BEFORE_LAST_BILLED_DATE));
+        }
+        if (dateTimeService.dateDiff(billingPeriodStartDate, lastBilledDate, false) >= 1) {
+            warningMessages.add(configurationService.getPropertyValueAsString(CUKFSKeyConstants.WARNING_CINV_BILLING_PERIOD_START_DATE_BEFORE_LAST_BILLED_DATE));
+        }
+
+        Pair<Date, Date> awardDateRange = parseDateRange(contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().getAwardDateRange());
+        if (dateTimeService.dateDiff(billingPeriodStartDate, awardDateRange.getLeft(), false) >= 1) {
+            warningMessages.add(configurationService.getPropertyValueAsString(CUKFSKeyConstants.WARNING_CINV_BILLING_PERIOD_START_DATE_BEFORE_AWARD_START_DATE));
+        }
+    }
+
+    protected ActionForward promptForFinalBillConfirmation(ActionMapping mapping, ActionForm form, HttpServletRequest request, HttpServletResponse response, String caller, String warningMessage, ContractsGrantsInvoiceDocument contractsGrantsInvoiceDocument) throws Exception {
+        ActionForward forward = null;
+
+        Object question = request.getParameter(KFSConstants.QUESTION_INST_ATTRIBUTE_NAME);
+        if (question == null) {
+            return performQuestionWithoutInput(mapping, form, request, response, CuArConstants.CINV_FINAL_BILL_INDICATOR_CONFIRMATION_QUESTION, warningMessage, KFSConstants.CONFIRMATION_QUESTION, caller, StringUtils.EMPTY);
+        }
+
+        Object buttonClicked = request.getParameter(KFSConstants.QUESTION_CLICKED_BUTTON);
+        if (CuArConstants.CINV_FINAL_BILL_INDICATOR_CONFIRMATION_QUESTION.equals(question)) {
+            if (ConfirmationQuestion.NO.equals(buttonClicked)) {
+                forward = mapping.findForward(KFSConstants.MAPPING_BASIC);
+            } else if (ConfirmationQuestion.YES.equals(buttonClicked) && contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().isFinalBillIndicator()) {
+                try {
+                    Pair<Date, Date> billingPeriod = parseDateRange(contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().getBillingPeriod());
+                    Date billingEndDate = billingPeriod.getRight();
+                    DateTimeService dateTimeService = SpringContext.getBean(DateTimeService.class);
+                    if (dateTimeService.dateDiff(billingEndDate, contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().getLastBilledDate(), false) != 0) {
+                        contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().setLastBilledDate(billingEndDate);
+                    }
+                } catch(Exception ex) {
+                    GlobalVariables.getMessageMap().putError(KFSConstants.GLOBAL_ERRORS, KFSKeyConstants.ERROR_CUSTOM, "promptForFinalBillConfirmation setting last billed date to end date: " + ex.getMessage());
```

Comment:
Now that this message is going to be displayed to the user, please update the wording accordingly to make it more user-friendly.

Also, should this code block return to the MAPPING_BASIC forward after adding the error to GlobalVariables?

Reply:
MessageMap.putError() will just print the referenced error message and parameters; it won't print the stack trace unless you explicitly include it (which would not be user-friendly). I would suggest adding a separate LOG.error() call to log more details about the exception and its stack trace.